### PR TITLE
Refactor calling optimization into DispatchOptimization.

### PR DESCRIPTION
**Pull request by ojura**
I noticed that @jihoonl opened the PR #726 which performs a similar thing. As discussed in googlecartographer/cartographer_ros#613 (@cschuet  has already taken a look), I pulled this out of #481 (a really old PR whose merging has been postponed), which is an example where re-running optimization is triggered from elsewhere as well (besides from `ComputeConstraintsForNode`). This refactoring makes libcartographer friendlier for use cases such as that one.

An important detail is that I have changed the condition in `WaitForAllComputations` to also check if the work queue is empty. If there are other things on the worker queue besides `ComputeConstraintsForNode`, currently we will wrongfully conclude that all computations are done. (This detail was merged in #754, so it's no longer in the diff of this PR).

Also missing is the same thing for 3D. I can add that when we settle on this.

Also, I suggest that `run_loop_closure` gets renamed to `run_optimization`.

### CODE REVIEW

**Comment by cschuet** [comment]
@wohe Wolfgang could you take a look at this?

**Comment by ojura** [comment]
Also, I think that changing the optimization problem options from within the worker queue, and dispatching optimization just like `ComputeConstraintsForNode` does is more consistent with the promise given in
https://github.com/googlecartographer/cartographer/blob/7994b9ac6b451d1de90dc9405f52c32c0f44f8d8/cartographer/mapping_2d/pose_graph.cc#L534-L535

Sure, if we first wait for all computations (and even if we do it properly by checking if the queue is empty, like in this PR), we can change the options in `optimization_problem_` without having to worry about the promise above. But what if _someone_ pushes _something_ to the worker queue during the final optimization? Chaos ensues (e.g. check in `WhenDone` fails). By making the final optimization a proper citizen of the worker queue, we don't have to worry about that at all.

File: cartographer/mapping_2d/pose_graph.cc [hunk 1]
```diff
@@ -506,14 +509,21 @@ void PoseGraph::AddTrimmer(std::unique_ptr<mapping::PoseGraphTrimmer> trimmer) {
...
+          options_.max_num_final_iterations());
+      DispatchOptimization();
+    });
+    AddWorkItem([this]() REQUIRES(mutex_) {
+      optimization_problem_.SetMaxNumIterations(
+          options_.optimization_problem_options()
+              .ceres_solver_options()
+              .max_num_iterations());
+    });
+  }
```
Comment (cschuet) [inline cartographer/mapping_2d/pose_graph.cc]: Can you explain the motivation for this change? Naively I would say that the below is more readable.

Comment (cschuet) [reply cartographer/mapping_2d/pose_graph.cc]: Oh, yes read your comment above which explains just that. So makes sense.

**Comment by ojura** [comment]
We can mirror this for 3D in a separate PR, or in this one, just let me know.

**Comment by ojura** [comment]
Updated.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 2.

**Pull request merged**